### DISCUSSION TITLE
disable uart txbuf

### DISCUSSION THREAD

**Discussion by Gaurav-Rajput01** [micropython/micropython#10941] [RP2040 / Pico]
want to disable tx buf of uart rp2040 micropython??
Anyone kindly help me ...

**Comment by robert-hh** [comment]
You could set it to a smaller value, like 32. But what do you want to achieve with that?

**Reply by Gaurav-Rajput01** [reply]
> You could set it to a smaller value, like 32. But what do you want to achieve with that?
first we need software handshaking (XON/ XOFF) but UART doesn't support or handle it. When we are receiving manually XOFF, Txbuffer already have more data for receiver, which occur problematic in our case.
we want to transfer character by character if possible.

**Reply by robert-hh** [reply]
The UART itself has not XON/XOFF flow control, and the SDK does not support it either. So you have implement it yourself by sending byte by byte and checking for XON/XOFF yourself. Sending bytes individually is possible, albeit not efficient. Using uart.txdone() you can check if supplied data has been sent. When txdone() returns true, still one character may be in the send process. So your receiver must be able to consume at least one more character after having sent XOFF. But that is usually the case.

Note: The UART has a 32 byte FIFO. Even when the buffering in the Software is disabled or set to minimum, the FIFO is still active and may hold data being sent.

**Reply by Gaurav-Rajput01** [reply]
Thanks for the response, I will try and let you know

**Comment by peterhinch** [comment]
The best way to do XON/XOFF flow control is usually to maintain your own receive buffer.

One task takes characters from the UART and puts them in the buffer and the main program removes them from the buffer. If the buffer holds more than `X` characters, XOFF is transmitted. When it holds fewer than `Y` characters, XON is transmitted. The size of the buffer `N`, and the value of `N - X`, is determined by the latency. Latency arises because of TX buffering and the time it takes for the remote device to respond to XOFF (the remote will have its own buffering).

To manage this concurrency you might consider using `uasyncio`.

**Reply by Gaurav-Rajput01** [reply]
Thanka for the response, I will try and let you know

**Reply by Gaurav-Rajput01** [reply]
@peterhinch @robert-hh when using uart.txdone() it shows error.The error is AttributeError: 'UART' object has no attribute 'tx_done'. I tried with lots of versions of micropython but got same error.

**Reply by robert-hh** [reply]
The name of the method is `txdone`, not `tx_done`. And you have to use a recent nightly build to get it. It was added August 2022.

**Reply by Gaurav-Rajput01** [reply]
@robert-hh I got same error with same txdone. And currently I tried with this version W5100S_EVB_PICO-20220618-v1.19.1 and 
W5100S_EVB_PICO-20230310-unstable-v1.19.1-963-g668a7bd28 in this version wiznet ethernet not connected.
If you have stable micropython version with wiznet kindly shre me the link.

**Reply by robert-hh** [reply]
What do you get when you run:
```
from machine import UART
print(dir(UART))
```
With the most recent nightly build `W5100S_EVB_PICO-20230310-unstable-v1.19.1-963-g668a7bd28 `, I get:
`['__class__', '__name__', 'any', 'read', 'readinto', 'readline', 'write', '__bases__', '__dict__', 'CTS', 'INV_RX', 'INV_TX', 'RTS', 'deinit', 'flush', 'init', 'sendbreak', 'txdone']
`
So the method txdone is present as method of the UART class. To use it, you may construct a wait loop like:
```
from machine import UART 
uart=UART(..whatever setting you need...)

# code that sends data

while not uart.txdone():
    idle()
```
Note that idle() may insert an delay of up to 1 ms. If that is too long, use instead either pass or a shorter sleep like time.sleep_us(100). After txdone() is finished, you may have to wait at the rp2 for another character time for the last character to be sent.

**Reply by Gaurav-Rajput01** [reply]
Thanks for your quick response @robert-hh now it's working if any problem occurs I will let you know
